Sources/TokamakDOM/Views/Canvas/Canvas.swift

```diff
@@ -0,0 +1,514 @@
+// Copyright 2020-2021 Tokamak contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//  Created by Carson Katri on 9/17/21.
+//
+
+import Foundation
+import JavaScriptKit
+import TokamakCore
+
+private let devicePixelRatio = JSObject.global.devicePixelRatio.number ?? 1
+
+private struct _Canvas<Symbols: View>: View {
+  let parent: Canvas<Symbols>
+  @State private var canvas: JSObject?
+  @Environment(\.inAnimatingTimelineView) private var inAnimatingTimelineView
+
+  var body: some View {
+    GeometryReader { proxy in
+      HTML("canvas", [
+        "style": "width: 100%; height: 100%;",
+        "width": "\(proxy.size.width * CGFloat(devicePixelRatio))",
+        "height": "\(proxy.size.height * CGFloat(devicePixelRatio))",
+      ])
+        ._domRef($canvas)
+        .onAppear(perform: draw)
+        ._onUpdate(perform: draw)
+    }
+  }
+
+  private func draw() {
```

Comment:
Would there be an easy way to split this into multiple functions to make it more readable?